pymatgen/analysis/diffraction/tem.py

```diff
@@ -0,0 +1,608 @@
+# coding: utf-8
+# Copyright (c) Pymatgen Development Team.
+# Distributed under the terms of the MIT License.
+# Credit to Dr. Shyue Ping Ong for the template of the calculator
+
+"""
+This module implements a TEM pattern calculator.
+"""
+
+from __future__ import division, print_function, unicode_literals
+import json
+import os
+from collections import namedtuple
+from fractions import Fraction
+from typing import List, Dict, Tuple, cast
+import numpy as np  # type: ignore
+import scipy.constants as sc  # type: ignore
+import pandas as pd  # type: ignore
+import plotly.graph_objs as go  # type: ignore
+import plotly.offline as poff  # type: ignore
+from IPython.display import set_matplotlib_formats  # type: ignore
+from pymatgen.core.structure import Structure  # type: ignore
+from pymatgen.analysis.diffraction.core import AbstractDiffractionPatternCalculator  # type: ignore
+from pymatgen.symmetry.analyzer import SpacegroupAnalyzer  # type: ignore
+
+with open(os.path.join(os.path.dirname(__file__),
+                       "atomic_scattering_params.json")) as f:
+    ATOMIC_SCATTERING_PARAMS = json.load(f)
+set_matplotlib_formats('retina')
+poff.init_notebook_mode(connected=True)
+
+__author__ = "Frank Wan, Jason Liang"
+__copyright__ = "Copyright 2020, The Materials Project"
+__version__ = "0.201"
+__maintainer__ = "Jason Liang"
+__email__ = "fwan@berkeley.edu, yhljason@berkeley.edu"
+__date__ = "02/30/2020"
+
+
+class TEMCalculator(AbstractDiffractionPatternCalculator):
+    """
+    Computes the TEM pattern of a crystal structure for multiple Laue zones.
+    Code partially inspired from XRD calculation implementation. X-ray factor to electron factor
+        conversion based on the International Table of Crystallography.
+    #TODO: Could add "number of iterations", "magnification", "critical value of beam",
+            "twin direction" for certain materials, "sample thickness", and "excitation error s"
+    """
+
+    def __init__(self, wavelength_cache: Dict[float, float] = None, symprec: float = None, voltage: float = 200,
+                 beam_direction: Tuple[int, int, int] = (0, 0, 1), camera_length: int = 160,
+                 debye_waller_factors: Dict[str, float] = None, cs: float = 1) -> None:
+        """
+        Initializes the TEM calculator with given voltage,
+        Args:
+            wavelength_cache (dict): Reference of voltage to relativistic wavelength. Value dependent on
```

Comment:
Let's actually get rid of this wavelength cache and use the `lru_cache` decorator from `functools` to take care of that caching.